### PR TITLE
Imp isolate eval

### PR DESCRIPTION
**Pull request by Leechael**

### CODE REVIEW

File: WapoJS/src/service.rs [hunk 1]
```diff
@@ -218,9 +263,37 @@ impl Service {
...
+        let ctx = self.context();
+        let default_fn = ctx.get_global_object().get_property("module")?.get_property("exports").unwrap_or_default();
         if default_fn.is_function() {
-            return self.call_function(default_fn, ());
+            let ret = self.call_function(default_fn, ());
+
+            //
+            // Auto set promise result to scriptOutput
+            //
+            if let Ok(val) = ret {
```
Comment (kvinwang) [inline WapoJS/src/service.rs]: Can we do this complex logic in JS?
for example define the following code in JS:
```js
Wapo.callModuleEntry = async function callModuleEntry() {
    const entry = globalThis.module?.exports;
    if (typeof entry=== 'function') {
         try {
               globalThis.scriptOutput = await entry();
          } catch (error) {
               globalThis.scriptError = error;
          }
    }
}
```
Then in the Rust we just do this:
```rust
let entry_fn = ctx.get_global_object().get_property("Wapo")?.get_property("callModuleEntry")?;
if entry_fn.is_function() {
    return self.call_function(entry_fn, ());
}
```

File: WapoJS/src/host_functions/isolate_eval.rs [hunk 2]
```diff
@@ -145,30 +144,38 @@ async fn wait_child(
...
         Some(output) if !output.is_undefined() => output,
         _ => output,
     };
+
+    let logs_obj = global_object.get_property("scriptLogs").ok();
+
+    let logs = match logs_obj {
+        Some(logs) if !logs.is_undefined() => logs,
+        _ => js::Value::Undefined,
+    };
```
Comment (kvinwang) [inline WapoJS/src/host_functions/isolate_eval.rs]:
```suggestion
    let logs = global_object.get_property("scriptLogs").unwrap_or_default();
```

**Comment by kvinwang** [comment]
Changes except the callModuleEntry part looks good to me.

File: WapoJS/src/service.rs (hunk 1 above)
Comment (Leechael) [reply WapoJS/src/service.rs]: ~~During my testing, I found that it became more difficult to obtain results from `isolateEval`. Therefore, I prefer to keep the current implementation.~~

I discovered that the issue was caused by a bug in handling the return value of `isolateEval`, which I have now fixed.

**Comment by Leechael** [comment]
Test output:

```
❯ for i in $(exec ls examples/commonjs/module/*.js); do echo $i; wapojs --worker-secret test $i; echo '\n\n'; done
examples/commonjs/module/guest-async.js
2024-09-06T10:37:43.593887Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:43.711261Z  INFO js::console: 'back to host'    
2024-09-06T10:37:43.712357Z  INFO js::console: 'the result of guest' { error: null, value: 'hello', logs: [ { message: '\'call in guest script\'', time: 1725619063606, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-async-returns-bigint.js
2024-09-06T10:37:43.762926Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:43.881011Z  INFO js::console: 'back to host'    
2024-09-06T10:37:43.882097Z  INFO js::console: 'the result of guest' { error: null, value: undefined, logs: [ { message: '\'call in guest script\'', time: 1725619063776, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-async-returns-boolean.js
2024-09-06T10:37:43.926893Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:44.045415Z  INFO js::console: 'back to host'    
2024-09-06T10:37:44.046501Z  INFO js::console: 'the result of guest' { error: null, value: false, logs: [ { message: '\'call in guest script\'', time: 1725619063940, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-async-returns-error.js
2024-09-06T10:37:44.091150Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:44.209154Z  INFO js::console: 'back to host'    
2024-09-06T10:37:44.210542Z  INFO js::console: 'the result of guest' { error: null, value: {}, logs: [ { message: '\'call in guest script\'', time: 1725619064104, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-async-returns-null.js
2024-09-06T10:37:44.269268Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:44.389087Z  INFO js::console: 'back to host'    
2024-09-06T10:37:44.390178Z  INFO js::console: 'the result of guest' { error: null, value: null, logs: [ { message: '\'call in guest script\'', time: 1725619064284, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-async-returns-number.js
2024-09-06T10:37:44.442763Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:44.561788Z  INFO js::console: 'back to host'    
2024-09-06T10:37:44.562849Z  INFO js::console: 'the result of guest' { error: null, value: 42, logs: [ { message: '\'call in guest script\'', time: 1725619064456, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-async-returns-plain-object.js
2024-09-06T10:37:44.608525Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:44.727126Z  INFO js::console: 'back to host'    
2024-09-06T10:37:44.728537Z  INFO js::console: 'the result of guest' { error: null, value: { runtime: 'wapo' }, logs: [ { message: '\'call in guest script\'', time: 1725619064621, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-async-returns-string.js
2024-09-06T10:37:44.775549Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:44.896173Z  INFO js::console: 'back to host'    
2024-09-06T10:37:44.897279Z  INFO js::console: 'the result of guest' { error: null, value: 'wapo', logs: [ { message: '\'call in guest script\'', time: 1725619064791, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-async-returns-symbol.js
2024-09-06T10:37:44.944155Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:45.062117Z  INFO js::console: 'back to host'    
2024-09-06T10:37:45.063211Z  INFO js::console: 'the result of guest' { error: null, value: undefined, logs: [ { message: '\'call in guest script\'', time: 1725619064957, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-async-returns-undefined.js
2024-09-06T10:37:45.107857Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:45.226634Z  INFO js::console: 'back to host'    
2024-09-06T10:37:45.227771Z  INFO js::console: 'the result of guest' { error: null, value: undefined, logs: [ { message: '\'call in guest script\'', time: 1725619065121, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/guest-sync.js
2024-09-06T10:37:45.281113Z  INFO js::console: 'call in host script'    
2024-09-06T10:37:45.297867Z  INFO js::console: 'back to host'    
2024-09-06T10:37:45.298084Z  INFO js::console: 'the result of guest' { error: null, value: 'hello', logs: [ { message: '\'call in guest script\'', time: 1725619065294, level: 2 } ], isError: false, isOk: true }    



examples/commonjs/module/host-async.js
2024-09-06T10:37:45.342317Z  INFO js::console: 'call in host script'    



examples/commonjs/module/host-sync.js
2024-09-06T10:37:45.374869Z  INFO js::console: 'call in host script'    
```

File: WapoJS/src/host_functions/isolate_eval.rs [hunk 3]
```diff
@@ -145,30 +144,46 @@ async fn wait_child(
...
+    let output_obj = global_object.get_property("scriptOutput").ok();
     let output = match output_obj {
         Some(output) if !output.is_undefined() => output,
-        _ => output,
+        _ => fallback,
     };
+
+    let serialized_obj = global_object.get_property("serializedScriptOutput").ok();
+    let serialized = match serialized_obj {
+        Some(serialized) if !serialized.is_undefined() => serialized,
```
Comment (kvinwang) [inline WapoJS/src/host_functions/isolate_eval.rs:159]: Passing JSValue from one js instance to another will cause memory corruption unless the value is `undefined` or `null`.

```suggestion
        Some(serialized) if !serialized.is_undefined() => serialized.to_string(),
```

**Comment by Leechael** [comment]
Move to new PR: https://github.com/Phala-Network/phat-quickjs/pull/26

**Pull request closed without merging**